src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector512.cs

```diff
@@ -0,0 +1,3107 @@
+        [Intrinsic]
+        [CLSCompliant(false)]
+        public static Vector512<uint> Shuffle(Vector512<uint> vector, Vector512<uint> indices)
+        {
+            Unsafe.SkipInit(out Vector512<uint> result);
+
+            for (int index = 0; index < Vector512<uint>.Count; index++)
+            {
+                uint selectedIndex = indices.GetElementUnsafe(index);
+                uint selectedValue = 0;
+
+                if (selectedIndex < Vector512<uint>.Count)
+                {
+                    selectedValue = vector.GetElementUnsafe((int)selectedIndex);
+                }
+                result.SetElementUnsafe(index, selectedValue);
+            }
+
+            return result;
+        }
+
+        /// <summary>Creates a new vector by selecting values from an input vector using a set of indices.</summary>
+        /// <param name="vector">The input vector from which values are selected.</param>
+        /// <param name="indices">The per-element indices used to select a value from <paramref name="vector" />.</param>
+        /// <returns>A new vector containing the values from <paramref name="vector" /> selected by the given <paramref name="indices" />.</returns>
+        [Intrinsic]
+        public static Vector512<float> Shuffle(Vector512<float> vector, Vector512<int> indices)
+        {
+            Unsafe.SkipInit(out Vector512<float> result);
+
+            for (int index = 0; index < Vector512<float>.Count; index++)
+            {
+                uint selectedIndex = (uint)indices.GetElementUnsafe(index);
+                float selectedValue = 0;
+
+                if (selectedIndex < Vector512<float>.Count)
+                {
+                    selectedValue = vector.GetElementUnsafe((int)selectedIndex);
+                }
+                result.SetElementUnsafe(index, selectedValue);
+            }
+
+            return result;
+        }
+
+        /// <summary>Creates a new vector by selecting values from an input vector using a set of indices.</summary>
+        /// <param name="vector">The input vector from which values are selected.</param>
+        /// <param name="indices">The per-element indices used to select a value from <paramref name="vector" />.</param>
+        /// <returns>A new vector containing the values from <paramref name="vector" /> selected by the given <paramref name="indices" />.</returns>
+        [Intrinsic]
+        public static Vector512<long> Shuffle(Vector512<long> vector, Vector512<long> indices)
+        {
+            Unsafe.SkipInit(out Vector512<long> result);
+
+            for (int index = 0; index < Vector512<long>.Count; index++)
+            {
+                ulong selectedIndex = (ulong)indices.GetElementUnsafe(index);
+                long selectedValue = 0;
+
+                if (selectedIndex < (uint)Vector512<long>.Count)
+                {
+                    selectedValue = vector.GetElementUnsafe((int)selectedIndex);
```

Comment:
Are we sure this cast won't overflow?

Reply:
Yes. `selectedIndex` is unsigned (so definitely positive) and the `if` validates it is less than the known positive `Vector512<long>.Count`, which will be `8`.